Add progress-projections dashboard

artdgn opened the pull request:
This adds a per country overview and projection of immunisation progress, ICU need and other data.

### Details:
- Data: progress projections is taking the overview table and adds:
  - Quantities per 100k population (by adding populations data),
  - Estimation of total cases from deaths,
  - Current immunisation percentage,
  - Current need for ICU beds per 100k population, and
  - Projection into future of immunisation percentages and ICU needs.
- Presentation:
  - Color-bars-table for current and future immunisation percentage by country - how far do
    we still have to go until this is over.
  - Color-bars-table current and future need for ICU beds per 100k - how bad is the
    medical crisis currently and in the future (unless growth rate
    changes) in each country.
  - Full overview table with estimation and projections for all countries and the base quantities (cases, deaths, etc).

Changes:
 - Refactor overview helper code. Only create helper module (without editing the original overview notebook).
 - Add overview template and country mappings into _data folder in the repo (to reduce external dependencies).
 - Add world population data file.
 - Add thumbnail for progress projections dashboard.

hamelsmu commented:
Some feedback


1. Run a grammar and spelling check on this notebook, there are many errors.

2. I'm confused by your use of `Immune Currently` 
   - the title `Immune Currently` suggests this number is a fact, not an estimation.  This is misleading
   - You don't discuss where this number comes from, or how it is modeled.
   - Even if you did have a model for this, I'm not sure how useful this number is 
   - Are you claiming that the 45% + of the entire country of Turkey is going to be immune in 60 days?  That is big claim, which perhaps deserves more illustration on how you arrived at this number


![image](https://user-images.githubusercontent.com/1483922/77856454-2e1a2380-71ac-11ea-9fd5-71cfc6562fd2.png)





The below is too busy, and confusing.  You need to reduce the information presented here.

![image](https://user-images.githubusercontent.com/1483922/77856420-d24f9a80-71ab-11ea-8b4f-24db57f003e6.png)


I'm afraid I cannot merge of fix this as is.  Tagging referees @zachmayer / @dansbecker for another opinion

artdgn commented:
Thanks for the feedback, will look into all of these and update.

dansbecker commented:
Interesting to see this... I have a couple quick thoughts

The phrase **Immune currently** makes it sound like there's a good understanding that recovering from infection bestows immunity. My impression from afar was that there's still some uncertainty about that. If I'm wrong, that's great to hear. If we're not sure, I wonder if a different word would convey less certainty on the issue... `Infected and Recovered` is wordy, and less compelling than `immune`. But it might be less speculative. Worth thinking about whether there's a better word for it.

I'm also surprised these rates are predicted by some model grow so quickly and then plateau

![image](https://user-images.githubusercontent.com/1390442/77873297-d0ff8b80-7206-11ea-87f7-d8fab6328944.png)


I'm guessing I've misinterpreted the columns, because it seems unlikely the number of people in the US who have been infected and recovered will go from 0.3% to 19% in two weeks... but if that does happen, I would have expected it to keep growing beyond 33% in the subsequent ~75 days. 

Also, it'd be nice to show a `%` sign in the reported fatality rate... Lastly, I'd put a label on case growth rate. Perhaps convert it to a percentage over some time period? But someone less familiar with the underlying calculation won't be sure what the units for that are.

artdgn commented:
I think I've addressed most of the feedback:
- Improved and fixed the growth rate and SIR calculations which cased the weird plateau effect earlier. Added visualisations to showcase for some counties (for illustration of calculations).
- Added a lot more clarifications in the presentations part. 

The two main points I didn't find a way to address, and just tried to clarify instead:
- The busy full table: I didn't find a way to balance better visual presentation and that table's purpose of enabling people to inspect a lot of details about their counties of interest. I've added a clarification of that purpose above it for now.
- "Immune" / "Immunisation" word. I didn't manage to find an alternative yet, but I've added clarifications of what is meant by it, and a lot of the assumptions.

dansbecker commented:
@artdgn Thanks. I find these numbers easier to understand with the plateau issue fixed.

I still have a question about how you interpret these (focusing on the top few results for simplicity).

![image](https://user-images.githubusercontent.com/1390442/77913425-763f5180-7251-11ea-832b-fa93a4ca70ba.png)

Do you think these are likely to be approximately right... or do you think of these just as the output of faithfully applying an SIR model, but likely not right? 

As a hypothetical to get at that question: if someone offered you a small bet about whether Spain will infect at least 75% of it's population in the next 60 days, would you take that bet because you trust the model enough... or would you think it's unlikely and still reject the bet?

artdgn commented:
> Do you think these are likely to be approximately right... or do you think of these just as the output of faithfully applying an SIR model, but likely not right?

@dansbecker interesting question. 
- The biggest assumption for the projection is the constant infection rate, which changes with policy and fear, but slowly. 
- I think the predictions are likely to be more right than wrong for the 14 days projections due to the momentum of the system. So I would take the 14 days bet.
- I don't think that in 60 or even 30 days the infection rate of any country in the top of that list will remain as it is today because of what is happening now and will happen soon (e.g. in 14 days). People will stop going out and getting infected at some point.

It's like a simple simulation of a driver not watching the road: the prediction in 14 meters is to go off the road, and in 60 meters is to go off a cliff. Once the driver looks up, he will slam the brakes, and swerve and the prediction will no longer be accurate. He may still avoid the cliff, but due to physics and the evidence for this guy not being a good driver, he's still very likely to be going off the road in 14 meters.

From the full table is looks like a lot of counties have more or less slammed the brakes already, but some are still busy changing the radio station.

Also, the model is quite simple, and there's still a lot that can be improved to make it more accurate. For example, to better model the recovery rate like in [Outstanding Cases](https://covid19dashboards.com/outstanding_cases/#Appendix:-Methodology-of-Predicting-Recovered-Cases).

On a broader note, I think the important part is the changing nature of projections with updated data, and that they will become more important as things develop. This is because the raw numbers are hard to interpret already. And at least for me, they don't answer the most important two questions in my mind:
1. How bad is it going to get?
2. When is it going to get better?
I'm sure that trained epidemiologists with better data are answering these questions to their local 
politicians daily. But I couldn't get these numbers anywhere.

artdgn commented:
- I've removed 60 day projection from main tables and added testing bias column to the first one, and reported "Immunity" to the second. 
- Also changed the order of the main tables to focus on the ICU need table first.
- I've also updated the way infection rate is calculated, the numbers are somewhat worse now for "top" countries.

dansbecker commented:
<p>Nice @artdgn. I feel more comfortable looking at the numbers without the long-term predictions. </p><p> </p><p>What are the units for <strong>Estimated testing bias from deaths</strong>?</p>

artdgn commented:
@dansbecker a couple more changes:
- Added more explanations and units to the testing ratio column. It's a ratio (so has no units), but I've made an effort to explain it better.
- I've updated the recovery calculation for to be like in [Outstanding Cases](https://covid19dashboards.com/outstanding_cases/#Appendix:-Methodology-of-Predicting-Recovered-Cases). The numbers have changed to a bit less pessimistic.

hamelsmu commented on _notebooks/2020-03-27-covid19-progress-projections.ipynb:
```diff
@@ -0,0 +1,2122 @@
+{
```
<p>Organize all the helper files into a folder under <code>_notebooks/</code> don't have a separate <code>_data</code> folder.&nbsp;&nbsp;This way I can easily see all the files related to this dashboard.   </p><p> </p><p>However, before you do all this work please review all of the feedback and see if this is something you want to do, as there is a fair amount of feedback here.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/github/covid19-dashboard/pull/232/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

artdgn replied:
<p>Thanks, done.</p>

hamelsmu commented on _notebooks/2020-03-27-covid19-progress-projections.ipynb:
```diff
@@ -0,0 +1,2122 @@
+{
```
<p>This question/answer format is confusing.  Give a very brief description in two sentences of what this dashboard is about.</p><p> </p><p>I don't understand what this statement is saying</p><blockquote>`Note: Assumption of current infection rates is not to say that they won't be reduced, but to understand the implications of not reducing them.</blockquote><p> </p><p>Consider deleting this and putting the general disclaimer we have on all dashboards with predictive models for consistency example:  <a href="https://covid19dashboards.com/covid-19-us-case-estimation/#Definition-Of-Fields:" target="_blank">https://covid19dashboards.com/covid-19-us-case-estimation/#Definition-Of-Fields:</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/github/covid19-dashboard/pull/232/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

artdgn replied:
<p>Changed completely, hope it helps.</p><p> </p>

hamelsmu commented on _notebooks/2020-03-27-covid19-progress-projections.ipynb:
```diff
@@ -0,0 +1,2122 @@
+{
```
<ul><li>Instead of <code>is done</code> say <code>calculated</code></li><li> Link to an appendix instead of saying details at bottom). </li><li>This section should clearly map to the fields in the table, and its not clear which fields match which definitions</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/github/covid19-dashboard/pull/232/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

artdgn replied:
<p>Done.</p>

hamelsmu commented on _notebooks/2020-03-27-covid19-progress-projections.ipynb:
```diff
@@ -0,0 +1,2122 @@
+{
```
<p>Define the fields in this table.  The testing bias ratio is not defined in this table. </p><p> </p><p>For example see "Definition of Fields" in this example: https://covid19dashboards.com/covid-19-us-case-estimation/#Definition-Of-Fields.&nbsp;Maybe you are already doing this but it is unclear which definitions go with which field.&nbsp;Make this explicit.&nbsp;</p><p> </p><p>This chart isn't that valuable without ICU capacity, and the comparison is not that useful, is there a way to include capacity in here so we know which countries are in trouble?</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/github/covid19-dashboard/pull/232/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

artdgn replied:
<p>I don't have per countries capacities for most countries. I've linked to what information I have more prominantly at the top of that section instead of only in appendix.</p><p> </p><p>Given the narrow range of ICU capacities in the link I've added, I think the comparative numbers (between countries, and between time ranges) are quite useful due to the differences in magnitude. </p><p> </p><p> </p><p>From just observing the past data - it seems that countries are in trouble earlier than their capacity is reached (probably because the trouble is localized), so to me it seems that any number above 1 is already in crisis mode (at least from countries that got into that territory (e.g. Italy).</p><p> </p><p>Of course, It shouldn't be difficult to add the capacity later if a data source is found (or if someone who has better ideas finds this). </p>

hamelsmu commented on _notebooks/2020-03-27-covid19-progress-projections.ipynb:
```diff
@@ -0,0 +1,2122 @@
+{
```
<p>Why is this table Top 20 by one metric, but sorted by another metric?  This is confusing. Based on the title you are filtering this for the top 20 countries by immunity but then sorting by total cases?   The total number of new cases is not even shown in this table.</p><p> </p><p>What am I supposed to take way from this table?  Turkey will really have 80% immunity in 30 days?  Where is that coming from?  Doesn't quite look right to me</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/github/covid19-dashboard/pull/232/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

artdgn replied:
<p>Added the sorting column and added clarifications.</p><p>Changed "Immunity" everywhere to "Affected percentage of population" to remove the speculative aspect of immunisation.</p><p> </p><p>The projection of Turkey heading to 80% in 30 days is based on the extremely high growth rate in recent days (above 30%). The model assumes constant infection rate. </p><p> </p><p>Although policy and fear will reduce the infection rate, I am not sure how we can model that reduction. </p><p> </p><p>Some counties have reacted quickly, some slowly, some have pre-empted the spread, and some are going exponential still. </p><p> </p><p>How would you propose to model the reduction in infection rate?</p><p> </p>

hamelsmu commented on _notebooks/2020-03-27-covid19-progress-projections.ipynb:
```diff
@@ -0,0 +1,2122 @@
+{
```
<p><code>is done from deaths by</code> needs to be <code>is calculated by</code></p><p> </p><p>Consider renaming this to the Appendix section</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/github/covid19-dashboard/pull/232/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

artdgn replied:
<p>Fixed and renamed.</p>

hamelsmu commented on _notebooks/2020-03-27-covid19-progress-projections.ipynb:
```diff
@@ -0,0 +1,2122 @@
+{
```
<p>Are these toy examples or is this the Actual model used for this Dashboard?  The wording / language could be clarified here.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/github/covid19-dashboard/pull/232/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

artdgn replied:
<p>These are the actual numbers. I've added clarifications.</p><p> </p>

artdgn commented:
@hamelsmu thanks for the detailed review and for considering my time as well.
The edits you're proposing seem to mostly be about clearer presentation / moving files around, and do not look like a lot of work. I have no problems with those. 
Are there any big issues regarding the content / approach itself that would prevent you from considering merging this aside from the presentation aspect?

hamelsmu commented:
@artdgn I think this dashboard would be very compelling if these points were addressed somehow:

- ICU maximum capacity of countries.  That way we know which countries are in trouble.
- I'm highly skeptical of the immunisation progress.  Why is Turkey 80%.  I would probably hide this section altogether because the numbers don't seem right to me, or at least I'm not convinced. 
    - If I were to get rid of this I would also want to get rid of immunisation in the entire dashboard. 

Let me know your thoughts.  Thanks again

artdgn commented:
@hamelsmu I've changed the presentation and the wording substantially and I think it should be better now. Most importantly, I've replaced "immunisation" with "affected population percentage", which should be much less contentious.

About the numbers not making sense. I've added some thoughts in the particular comment. I will be happy to hear any suggestions to make the projections more accurate (and I have some ideas I'd like to play around). 

But in general:
>  ..the numbers don't seem right to me

Would today's numbers seem right to you 30 days ago? Would last month's numbers make sense two months ago?

hamelsmu commented on _notebooks/2020-03-27-covid19-progress-projections.ipynb:
```diff
@@ -0,0 +1,2126 @@
+{
```
<p> </p><p> </p><ul><li><code>ICU capacities differ by country but seem to be in the range of&nbsp;<a href="https://www.forbes.com/sites/niallmccarthy/2020/03/12/the-countries-with-the-most-critical-care-beds-per-capita-infographic/" target="_blank">2-10 for developed countries</a>.</code> I linked to a wikipedia article where you can find this data in another comment and add that into this dashboard, so you can remove this bullet.</li><li>This definition is ambiguous - how is this ratio calculated? Can we provide a more precise calculation?  <code><strong>Estimated testing bias ratio from deaths</strong>: estimated ratio of how reduced is the testing coverage relative to countries which performed comprehensive testing.  </code></li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/github/covid19-dashboard/pull/232/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

hamelsmu replied:
<ul><li>"New Deaths in 5 days" needs to be "New Deaths in last 5 days"</li><li>I think you can add ICU capacity into this chart. I Googled "ICU capacity by country" and<a href="https://en.wikipedia.org/wiki/List_of_countries_by_hospital_beds" target="_blank"> this wikipedia article</a> was the third result, which has a table of data. Instead of ICU capacity per 100k, suggest showing "Estimated percent of ICU capacity..."</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/github/covid19-dashboard/pull/232/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

hamelsmu replied:
<ul><li>"reported total deaths to total cases percentage" needs to be "reported total deaths divided by total cases."</li><li>The field "<strong>Estimated new cases in 5 days</strong>." doesn't have a definition.  Also should be "in last 5 days" not "in 5 days"</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/github/covid19-dashboard/pull/232/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

artdgn commented:
@hamelsmu 
ICU:
- Thanks for that. The wikipedia table only had information for 36 counties in this dashboard (OECD mostly) and I've added some more data from a paper about ICU beds in Asian countries (form a paper that was cited in that Wikipedia page). 
- I've added this data into the first table, and for most countries in the top of the list the data is available. However, I am not calculating the ratio (of need / capacity), because the ICU data is only available for a minority of the countries (30% - 52 / 172). 
- I have also added a rough estimation of available spare capacity (a large percentage of the ICU beds are occupied normally) - which I think is the more important number to compare to. I've added sources to support that estimation.

Testing bias:
- I've removed it both due to the difficulty of explaining it and due to adding the ICU data.

hamelsmu commented:
@artdgn thank you so much for slogging through these reviews and making these changes, your perseverance is appreciated.  I think this is almost ready.  One last change.  I believe the most important metric in this whole dashboard is

![image](https://user-images.githubusercontent.com/1483922/78174421-1346e980-740e-11ea-8dff-9b68dab0352a.png)

Would you agree?  

- Since the data is available for most of the countries in the top of the table, and that's the only part of the table we are showing -- why not calculate the `Estimated % of spare ICU capacity needed`  Is it because you are afraid the table might change or the sorting order will eventually change?

- Perhaps switch all the `projected` columns to show this number?  I understand that 3 out of the top 20 countries will not have metrics.  But I think your dashboard has the most impact highlighting this information.

hamelsmu commented:
Sorry I merged this by accident.... Please reopen this PR when ready, I reverted the merge

artdgn commented:
@hamelsmu thanks for the encouragement, and I appreciate the time you're dedicating to these reviews and to maintaining the project. It's great that you're insisting on quality, if this project should help anyone make more informed decisions, it should be the best data, explained in the most clear and trustworthy way.

Initially I've tried making these changes, and I've tried finding / inputing the missing data. However:
- I couldn't find much more data for the missing countries. It would have to be googling country by country and adding manually (for those that can be found).
- I've thought to impute the data based number of regular beds (for which there are good sources), or by using socio-economic status, or just continent. The problem is that all of these don't correlate well to the data that I do have, and would be mostly wrong. So I've abandoned that attempt. 

Also, the sorting is a big concern. That table is now sorted by the added ICU need (which is estimated similarly for all countries), but if we impute ICU capacity, and sort by the new "ratio to spare capacity", that sorting is going to be different and much more arbitrary (due to the imputation), and imputed countries are going to land in arbitrary places.

I've also considered splitting the table into two tables - countries that have ICU data and those that don't. We'll naturally present the "haves" table first, and this doesn't seem right to me, as this is also an arbitrary difference, especially since the "haves" are also the richest countries (OECD).

But perhaps most importantly, right now it's quite easy to grasp that relationship visually - because the bars (the red ones for need, and the blue ones for capacity) are on the same scale (0-10), so at least when one of the numbers is below 10 it's easy to grasp where red is bigger than blue or otherwise. This conveys this "ratio" information in a natural way, but perhaps needs to be clarified in the presentation. 

So what I did for now (except the various data and imputaions which I didn't add) is add a clarification about the bars being on similar scale for easier comparison.

This issue seems a bit tricky to solve, but I don't think it makes the dashboard not useful comletely. 
Perhaps it's worth to merge it as it is now, as it is already in a much better shape than a few days ago, and iterate on the possible improvements later. 

I think as in many cases, it's better to release and iterate on actual feedback from users than to keep refining without it. (Btw is there a place for comments and feedback other than the issues tab?)
